File: ruby/matching-brackets/matching_brackets.rb
class Brackets
  MAP = {
    '[' => ']',
    '(' => ')',
    '{' => '}'
  }.freeze

  def self.paired?(str)
    stk = []
    str.chars.each {
      MAP[_1] ? stk << MAP[_1] : MAP.value?(_1) &&
        (return false if stk.empty? || stk.pop != _1)
    }
    stk.empty?
  end
end

File: cpp/matching-brackets/matching_brackets.cpp
#include "matching_brackets.h"
#include <stack>
#include <unordered_set>
#include <map>

namespace matching_brackets {
    static const auto closing_brackets = unordered_set<char>{'}', ']', ')'};
    static const auto opening_brackets = map<char, char>
            {{'{', '}'},
             {'[', ']'},
             {'(', ')'}};

    auto check(const string &s) -> bool {
        auto stk = stack<char>{};
        for (auto c: s) {
            if (opening_brackets.find(c) != end(opening_brackets)) {
                stk.push(c);
            } else if (closing_brackets.find(c) != end(closing_brackets)) {
                if (stk.empty() || c != opening_brackets.at(stk.top())) return false;
                stk.pop();
            }
        }
        return stk.empty();
    }

}
